kong/plugins/prometheus/exporter.lua

```diff
@@ -168,6 +174,50 @@ local function collect()
                  "/metrics endpoint: ", err)
   end
 
+  -- upstream targets accessible?
+  local balancer = require "kong.runloop.balancer"
+  local upstreams_dict = balancer.get_all_upstreams()
+  for upstream_name, upstream_id in pairs(upstreams_dict) do
+    -- same logic from kong.db.dao.targets
+    local health_info
+    health_info, err = balancer.get_upstream_health(upstream_id)
+    if err then
+      ngx.log(ngx.ERR, "failed getting upstream health: ", err)
+    end
+
+    if health_info then
+      local status = 0 -- healthchecks is off
+      for target_name, target_info in pairs(health_info) do
+        if target_info ~= nil and
+          #target_info.addresses > 0 then
+          for _, address in ipairs(target_info.addresses) do
```

Comment:
Why is health check info not provided for each address?

Reply:
this code is a copy from kong.db.dao.targets, it would be nice if the dao.targets provides it right the way

Reply:
@hbagdi here is an example of the `health_info` table:

```lua
  ["127.0.0.1:8001"] = {
    addresses = {
      {
        health = "UNHEALTHY",
        ip = "127.0.0.1",
        port = 8001,
        weight = 50
      }
    },
    host = "127.0.0.1",
    nodeWeight = 50,
    port = 8001,
    weight = {
      available = 0,
      total = 50,
      unavailable = 50
    }
  },
  ["httpbin.org:80"] = {
    addresses = {
      {
        health = "HEALTHY",
        ip = "34.198.151.234",
        port = 80,
        weight = 1000
      },
      {
        health = "HEALTHY",
        ip = "52.6.108.56",
        port = 80,
        weight = 1000
      }
    },
    host = "httpbin.org",
    nodeWeight = 1000,
    port = 80,
    weight = {
      available = 2000,
      total = 2000,
      unavailable = 0
    }
  }
}
```

Reply:
I was going to hit merge but there is a problem we need to fix.
With the current code, the output is:
```
kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",state="dns_error"} 0
kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",state="healthy"} 0
kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",state="unhealthy"} 1
```

This should instead be a per address output, where one address can be healthy while the other one can be unhealthy:
```
```
kong_upstream_target_health{upstream="foo.com",target="52.6.108.56:80",state="dns_error"} 0
kong_upstream_target_health{upstream="foo.com",target="52.6.108.56:80",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="foo.com",target="52.6.108.56:80",state="healthy"} 0
kong_upstream_target_health{upstream="foo.com",target="52.6.108.56:80",state="unhealthy"} 1
kong_upstream_target_health{upstream="foo.com",target="34.198.151.234:80",state="dns_error"} 0
kong_upstream_target_health{upstream="foo.com",target="34.198.151.234:80",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="foo.com",target="34.198.151.234:80",state="healthy"} 0
kong_upstream_target_health{upstream="foo.com",target="34.198.151.234:80",state="unhealthy"} 1
```

Reply:
@hbagdi the tests cover it. At the tests I have one target healthy, one unhealthy and one with DNS Error for the same upstream. Also I have one upstream without healthchecks.
Could you detail a little bit more how do you get this scenario ?

Reply:
@hbagdi I just ran kong here and:
```
kong_upstream_target_health{upstream="upstream1",target="httpbin.org:80",state="dns_error"} 0
kong_upstream_target_health{upstream="upstream1",target="httpbin.org:80",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="upstream1",target="httpbin.org:80",state="healthy"} 1
kong_upstream_target_health{upstream="upstream1",target="httpbin.org:80",state="unhealthy"} 0
kong_upstream_target_health{upstream="upstream1",target="postman-echo.com:80",state="dns_error"} 0
kong_upstream_target_health{upstream="upstream1",target="postman-echo.com:80",state="healthchecks_off"} 0
kong_upstream_target_health{upstream="upstream1",target="postman-echo.com:80",state="healthy"} 0
kong_upstream_target_health{upstream="upstream1",target="postman-echo.com:80",state="unhealthy"} 1
```

Reply:
@hbagdi yes! One "IP+PORT" can be unhealthy and other can be healthy, but Kong itself consider the target as "Healthy" when calling `http://admin_api/upstreams/:upstream/targets/health` once it uses the code from `kong.db.dao.targets` to get this information.
In your opinion, should we add "address" state too? 
```
kong_upstream_target_health{upstream="foo.com",target="httpbin:80",address="34.198.151.234:80",state="healthy"} 1
```

Reply:
How does kong work? If I have one `address` of a `target` unhealthy does kong route to it?

Reply:
What I understand is that a DNS name as a target, in our example here, `httpbin.org`, is not technically even a target. In this case, it resolves to two A records and the targets then are the two IP addresses (keeping port out of the discussion currently). Kong then performs health check for each of the IP address. Each IP address is a target and can be healthy or unhealthy in this case. 

@hishamhm Please correct me if I'm wrong.

Reply:
Once the address of the target is considered UnHealthy kong stops sending traffic to that host, right?
Until all hosts are consired UnHealthy, the target keeps with "Healthy" status (that is what I think by reading the code).

Reply:
Your point there is still valid. Do we need to expose the healthiness of each address from the targets from the upstreams?

Reply:
> Do we need to expose the healthiness of each address from the targets from the upstreams?

Yes. That's what makes most sense.

I also question if we should have target and address as two different labels or not. Should we only have a target label, with the value of "address"?

@Tieske ping (in case you have anything to add).

Reply:
In my opinion, it makes easy to identify a possible downtime easily when I have the upstream name. 
Imagine I have these two targets: `internal-afaa16a090c7311ea802a02957de333b-5186484.sa-east-1.elb.amazonaws.com` and `internal-a9580cd0a0a2211eaa3a30a8dbebae10-354287126.sa-east-1.elb.amazonaws.com` it is really difficult to figure it out from each upstreams it belongs to.
I can use the label to send clearly notifications to my NOC team.

Reply:
Shall I change this PR to expose { `upstream_name` `target_name:target_port` `address_ip:address_port` `health_state` } ?

Reply:
Yes, I suspected that would be a good case.
I wish we could avoid this to limit the impact on performance but that seems to be very limiting. Let's then have a `target` and `address` labels in that case. `target` is what user configures, i.e. `httpbin.org`. `address` is the resolved IP address and port (from DNS record (in case of `SRV`) or port from the target).

Reply:
Yes. An example would be then:
```
kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",address="34.198.151.234:80",state="healthy"} 1
```

Reply:
> Yes. An example would be then:
> 
> ```
> kong_upstream_target_health{upstream="foo.com",target="httpbin.org:80",address="34.198.151.234:80",state="healthy"} 1
> ```

LGTM, I'm changing it here

Reply:
hmm in case of "DNS ERROR" I wont have addresses, how should we populate this scenario? "null"?

Reply:
Yes, let's make the `address` label null in that case, that should remove the label from the output or output an empty string, either of which should be okay.

Reply:
It should be ok now